### ISSUE TITLE
OpenJDK JVM deadlock triggered with jemalloc 5.x?

### ISSUE THREAD

**Issue by djscholl** [jemalloc/jemalloc#1392]
We've found an issue that we can only reproduce when LD_PRELOADing jemalloc 5.1. The issue causes a deadlock where one or more threads are waiting to lock an object monitor that no thread is currently holding. When we attempted to debug, we discovered the thread that is expected to be holding the lock has instead left the synchronized block and has returned to the thread pool. If we don't LD_PRELOAD jemalloc and rely on glibc malloc we're unable to reproduce the issue. If we use jemalloc 4.5.0 we're unable to reproduce it there as well.

We've written a simple test application that can reproduce the issue.

Test application - https://github.com/djscholl/jemalloc-java-deadlock
Example log output when a deadlock occurred - https://gist.github.com/djscholl/413071ef29671fb53b5a64e105421f1a

In the example log output, we can see that 999 threads are
```
        -  blocked on sun.security.provider.Sun@65e2dbf3
```
Which is supposedly owned by "pool-1-thread-558", but that thread doesn't appear to be holding the monitor.
```
"pool-1-thread-558" Id=566 RUNNABLE
        at java.lang.Class.getConstructor(Class.java:1825)
        at java.security.Provider$Service.newInstance(Provider.java:1594)
        at sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
        at sun.security.jca.GetInstance.getInstance(GetInstance.java:164)
        at java.security.Security.getImpl(Security.java:695)
        at java.security.MessageDigest.getInstance(MessageDigest.java:167)
        at com.example.DeadlockTest.lambda$null$1(DeadlockTest.java:44)
        at com.example.DeadlockTest$$Lambda$7/2093176254.run(Unknown Source)
        ...

        Number of locked synchronizers = 1
        - java.util.concurrent.ThreadPoolExecutor$Worker@2ff4f00f
```

We've reproduced this using a jemalloc.so built with everything set to the default and no malloc options set. We've tried it with --enable-debug and can reproduce it with that as well. We've tried with background threads enabled and disabled, as well as changing number of arenas. All of these configurations still result in reproducing this issue.

We've tried this on a few different OpenJDK versions as well.
- Multiple versions of 1.8 between 1.8.0.121 and 1.8.0.202
- 1.9
- 1.11

It's a little awkward to post a JVM dead lock to jemalloc developers, but we're stuck! We'll post this same issue on OpenJDK and link that here when we have it. 

We tried running the test application with `-DwaitOnDeadLock=true` and attached gdb, and all we see is that the threads are in `pthread_cond_wait` while the thread ID that they're waiting for is elsewhere outside of the synchronized block (so it shouldn't be holding the object monitor any more).

**Comment by interwq** [comment]
Thanks for the detailed report! I believe we have seen the same issue with our Java developers as well. The repro case you shared should help us to investigate; but to be honest I'm no expert on JVM so not sure how much I can help.

Since you mentioned that it doesn't repro with jemalloc 4; taking a wild guess that it has something to do with our thread local usage in jemalloc 5 (we inlined the thread cache in to thread local storage with 5.0). Can you please try build jemalloc with `--disable-initial-exec-tls`? See the details in https://github.com/jemalloc/jemalloc/blob/dev/INSTALL.md

**Comment by djscholl** [comment]
Thanks for the quick suggestion - it seems to have solved the problem! We're going to do some additional testing to make sure.

**Comment by interwq** [comment]
Thank you for sharing all the info @djscholl! If it works we would still need to figure out why, but likely it's going to an issue on the JVM's side.

**Comment by djscholl** [comment]
I agree that this needs to be understood better, but I don't think there's much more we can do on our end to root cause it. I reported an issue on the JVM as well but don't yet have a publicly visible issue ID. I'll post back if I get one.

**Comment by djscholl** [comment]
The Java Bug Database confirmed that they can reproduce this on Java 12, and opened a bug for us - https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8215355

**Comment by interwq** [comment]
@djscholl did you get a chance to test this more to confirm that `--disable-initial-exec-tls` solves the issue? I have a colleague testing it in their set up too; but would be nice to confirm with you guys as well.

**Comment by djscholl** [comment]
It's still looking good on our end. We're working on bringing this change to our application to get some test runtime there, including performance testing.

**Comment by djscholl** [comment]
Actually bad news... we are able to reproduce it even with your suggestion on 1.8.0_91. Going to do some testing on the newer JVM versions.

**Comment by djscholl** [comment]
Hi @interwq - can you take a look at the latest comments on the openjdk issue? They're pretty convinced that this is a jemalloc problem. Has anyone on your side looked more closely at 498856f44a30b31fe713a18eb2fc7c6ecf3a9f63?

We tried a couple more jemalloc configurations with no luck - it still reproduces.

./configure --enable-stats --enable-fill  --with-malloc-conf=junk:true
./configure --enable-stats --enable-fill  --with-malloc-conf=zero:true

The JVM devs mentioned that it doesn't appear to be reproducible on a debug build which does extra memory zapping. I'm not sure exactly what they mean by that.

**Comment by interwq** [comment]
Thanks for the info @djscholl. I looked at the commit but it was a rather large refactoring on an internal core data structure. I'll dig into it and post here about the findings.

**Comment by davidtgoldblatt** [comment]
One configuration that might be worth testing is `--enable-debug`, if it reproduces there.

**Comment by davidtgoldblatt** [comment]
Oh I'm sorry, I see you already tested it. Please disregard.

**Comment by djscholl** [comment]
Any updates from the jemalloc side? Doesn't seem like the jdk side has hit any breakthroughs. Do you guys maybe have a feature branch that we could use to break down 498856f44a30b31fe713a18eb2fc7c6ecf3a9f63 into smaller chunks so we can more easily identify the problematic code change?

**Comment by interwq** [comment]
Spent some time digging but nothing conclusive yet. We have some year-end review stuff to do. It will take another week or so before I can come back to this.

**Comment by jasone** [comment]
Re: bisecting to 498856f44a30b31fe713a18eb2fc7c6ecf3a9f63, note that this change was part of a ~50-step rewrite of the core data structures and algorithms.  As such, it is probably best not to dwell on the specific revision, since there were inevitably numerous regressions introduced that we found and fixed later on.

**Comment by davidtgoldblatt** [comment]
The fact that it doesn't show up with biased locking disabled (from the openjdk bug) is interesting. This probably leads to some heavy fences being inserted on native/managed transitions. On older versions of glibc (before 2.26), every allocation has (some operation equivalent to) a heavy fence down an allocation path. So the allocator could be hiding missing synchronization.

Out of curiousity, what glibc are you using? If it's before 2.26, it would be interesting to see what happens if you try with a version after 2.26 (which introduced thread-local caching, and therefore no longer does a heavy fence per allocation).

**Comment by davidtgoldblatt** [comment]
Actually, an easier way to test that would be to insert `atomic_fence(ATOMIC_SEQ_CST)` into the jemalloc entry points and see if the deadlocking goes away. That would be fairly definitive I think.

**Comment by davidtgoldblatt** [comment]
And you could get a good approximation by disabling the jemalloc tcache, too, by setting "tcache:false" in the malloc conf; you wouldn't even have to rebuild.

**Comment by interwq** [comment]
The JDK locking bug in https://bugs.openjdk.java.net/browse/JDK-8213825 was fixed recently -- @djscholl can you help to test and see if it repros with JDK12 b30 (the build with the locking fix)?

**Comment by ajbaldocchi** [comment]
I've run the test case with OpenJDK 12b30 and it definitely still reproduces.

**Comment by interwq** [comment]
Thanks for help testing @ajbaldocchi. I'll try digging more then.

**Comment by djscholl** [comment]
Learn anything new on this @interwq? You said your java developers see this as well, right? Do they have any workaround for it?

**Comment by interwq** [comment]
@djscholl : unfortunately I haven't had any luck on this. It appears that some of our java binaries suffer from this while others don't; so at the moment we haven't switched our internal JVM to jemalloc completely.

In the OpenJDK thread, it was mentioned that BiasedLocking might be related, can you try if _enabling_ it workarounds the issue? (the thread suggests disabling it, but just in case). It still looks to be a concurrency bug outside of jemalloc somewhere.

**Comment by jestan** [comment]
We are using rocksdb java with jemalloc for performance reasons and now thinking of reverting to 4.5 or glibc, the severity of the issue is scary.

It seems OpenJDK developers are closed this issue saying that  "All appearances are that it is a jemalloc issue related to the thread caching. Closing as "External" "

And as others have said, this issue is not reproducible after disabling tcache in jemalloc

```bash
export MALLOC_CONF=tcache:false
```

(For all below tests I have used Jemalloc 5.2.0 and Java 1.8.0_201)

What I have found that this issue quickly reproduces by using following snippet, the original reproducer stops after 30th iteration but with following change it not even finishing the first iteration

```java 
            synchronized (lock) {
              ThreadLocalRandom.current().nextLong();
              myAdder.increment();
            }
```
It seems this issue happens with only with accessing thread local with in a synchronization block ?
the thread local random is implemented by accessing thread field using Unsafe.put and get

So i tried to reduce this issue by copying some part of ThreadLocalRandom Unsafe logic to a minimal reproducer shown below and got it reproduced after 13th iteration

```java
          synchronized (lock) {
              UNSAFE.putInt(Thread.currentThread(), PROBE, 0);
              myAdder.increment();
            }
```

And I also tried to reproduce with a non thread local Unsafe access like below but the issue was not reproducible 

```java

           synchronized (lock) {
              UNSAFE.getInt(DATA, VALUE);
              myAdder.increment();
            }
```

I dont have much idea how this Unsafe works :) , It uses additional load store fences ?
http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/tip/src/share/vm/prims/unsafe.cpp

**Comment by interwq** [comment]
Looking into this again now. Hopefully will have something tomorrow.

**Comment by interwq** [comment]
Some updates:
- `tcache:false` only makes the issue happen way less frequently. It still repros for me after a while. Without tcache the timing is quite different (locks involved on every malloc / free).
- Narrowed it down to [this allocation when jvm is starting a new thread](https://github.com/unofficial-openjdk/openjdk/blob/a164493c87365b836c1fa65ebdd6899f24f1bb91/src/hotspot/share/prims/jvm.cpp#L2815). I'm quite certain that, with padding pages before and after the allocated memory, it doesn't happen anymore (looped the test 10k times).

The specific allocation has `size == 3008`; with the stack trace below --

```
#0  malloc (size=<optimized out>) at src/jemalloc.c:2388
#1  0x00007ffff5d1d665 in os::malloc(unsigned long, MemoryType, NativeCallStack const&) ()
   from /usr/local/fbprojects/packages/fb-jdk-8/12/jre/lib/amd64/server/libjvm.so
#2  0x00007ffff5e86da4 in Thread::allocate(unsigned long, bool, MemoryType) ()
   from /usr/local/fbprojects/packages/fb-jdk-8/12/jre/lib/amd64/server/libjvm.so
#3  0x00007ffff5b110c0 in JVM_StartThread ()
   from /usr/local/fbprojects/packages/fb-jdk-8/12/jre/lib/amd64/server/libjvm.so
#4  0x00007fffe40bf7a7 in ?? ()
#5  0xfffffffe00000000 in ?? ()
#6  0x00007fffe40bf4c2 in ?? ()
#7  0x00007ffff7fdcc00 in ?? ()
#8  0x00007fff99c2e628 in ?? ()
#9  0x00007ffff7fdcc60 in ?? ()
#10 0x00007fff99cb63e8 in ?? ()
#11 0x0000000000000000 in ?? ()
```
Some sort of stack overflow could explain this. I'm adding mprotect on the padding pages to see if it catches anything. Will pass this back to the JDK developers if nothing else in the next couple of days.

**Comment by davidtgoldblatt** [comment]
How'd you narrow it down there?

**Comment by interwq** [comment]
@davidtgoldblatt : what helped most was, bisecting the specific size range to add padding. Also found that with narenas:1 it repros faster which made the bisecting easier.

**Comment by interwq** [comment]
Based on the information gathered, this doesn't look like a jemalloc bug. 

Summary of the findings:
- With [our most recent production commit](https://github.com/jemalloc/jemalloc/commit/4fbbc817c1130d3d6c066f132fb5a2b23803be89), setting `narenas:1` makes the test program deadlock faster (usually ~30 runs or less) for me.
- The issue still repros with `tcache:false` but does take longer. Without thread cache, every malloc/free involves locking which affects the timing; in addition memory layout change as well.
- The minimal change to make the deadlock go away, is to pad at the beginning of this allocation for [creating new JavaThread](https://github.com/unofficial-openjdk/openjdk/blob/8125e195c884796eb5e766cdd2151c7aff80b7a6/src/hotspot/share/runtime/thread.cpp#L178). On my server the input size is 968 w/ alignment 2K (`UseBiasedLocking == true`), resulting the malloc size 3008 (this falls into the 3072-byte size class). Shifting the returned pointer by 16 byte on `malloc(3008)` (and restore the shift on free accordingly) make the deadlock disappear.
- Also tried padding after the actual allocation, which only makes the deadlock less frequent. Still repros even with a 1M padding after. 
- For the 3072 size class, jemalloc uses 3-page extents, so each 12K extent contains 4*3072 allocations. Therefore all 3072 allocations are 1024 aligned by default already.
- Because of the [align_up](https://github.com/unofficial-openjdk/openjdk/blob/8125e195c884796eb5e766cdd2151c7aff80b7a6/src/hotspot/share/runtime/thread.cpp#L181), bumping 16 byte can make JVM to shift the actual JavaThread object by 2K, e.g. for the first item in the 12k extent which was 4K aligned (before the padding).
- Allocating all 3008  allocations from a separate pool of memory (still using 3072-byte to fulfill) also stops  the deadlock, without the help of any padding. (I believe jemalloc 4.x is leaning towards that behavior, which is why it doesn't happen there). This means the deadlock only happens when the JavaThread objects can live close to some other allocations.
- To further check the neighbor requirement, I tried only pad the first item in a 12K extent (which is the only JavaThread that can live right next to other allocations). This also makes the deadlock go away.

12K extent: 
0---1K---2K---3K---(4K)---5K---6K---7K---(8K)---9K---10K---11K---(12K)
A--------------B-----------------C-----------------D--------------------

i.e. only shift the return address of the A allocation above alone by 16B "fixed" the deadlock. Then that JavaThread object will be aligned at 2K instead of at 0, which makes all JavaThreads at least 1K away from other allocations. Looped the test program 10K times and no repro. This is the [test commit to add padding](https://github.com/interwq/jemalloc/commit/ccd065fd3c65a6e0f84d1114df7165f4986127eb#diff-d2be2656d37670b0c7f5f4320c0670ceR2408)

- Also tried to pad a 4K page before the JavaThreads and `mprotect`, but didn't catch any violations. And the deadlock also goes away with the 4K padding.
- Tried writing some magic words in the 16B padding, or even the bumped spare space from `align_up`; didn't observe any change in them.

To sum up,  something could be happening when the JavaThread object is right next to some other objects (wild guess, Biased Locks?); but no directly corruption observed AFAICT. Something tricky with the pointer diff is < 1K with biased locks?

**Comment by interwq** [comment]
I can't seem to post to the JDK issue https://bugs.openjdk.java.net/browse/JDK-8215355. Can someone please help to share the info above to the JDK guys? Thanks. https://github.com/jemalloc/jemalloc/issues/1392#issuecomment-549554636
@djscholl, @jestan, @amir-shalem, @ajbaldocchi

**Comment by djscholl** [comment]
Thanks for digging into this @interwq! Looks like some solid evidence towards this being a JVM issue. I've sent an email to Fairoz Matte to ask him to reopen it. I'm also unable to comment directly on the JDK issue.

**Comment by interwq** [comment]
Looks the bug was fixed in JDK below. Thanks to everyone helped on the investigation!

https://bugs.openjdk.java.net/browse/JDK-8215355

**Comment by interwq** [comment]
@djscholl -- sorry need to bother you again (still couldn't find a way to post on the OpenJDK thread directly): could you please help ask if the JDK developers will consider back porting the fix to earlier JDK versions 8 and 11? We had multiple users asking as they couldn't upgrade jemalloc and it would be great to enable jemalloc for them too.

**Comment by djscholl** [comment]
Will do! Thanks so much for digging into this @interwq!

**Comment by djscholl** [comment]
Fairoz was on vacation for a bit but I got a response from him today. He says: "We are planning to backport this to 11 and 8, but it may take some time to get that into the production builds."

**Comment by interwq** [comment]
Sounds great. Thanks again Dan!

**Comment by LiuDui** [comment]
@djscholl Sorry to interrupt. I have encountered a similar problem recently several times: Object monitor deadlock with no threads holding the monitor. I am using jre1.8.0_181 with jemalloc5.2 on centos7.  

```
java -version
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)
```

I checked the code and found that the fix was not incorporated. But I cannot reproduce it on centos with test application https://github.com/djscholl/jemalloc-java-deadlock. Is this problem related to the operating system? Or is this problem only in openJDK? I'm a little confused.

**Comment by ajbaldocchi** [comment]
@LiuDui The jemalloc-java-deadlock test attempts to make a rare condition more common by running many many iterations, but failure to reproduce does not guarantee the issue is not present in your current Java version as it is quite timing-dependent. You are on a build that is vulnerable to JDK-8215355, the fix of which was backported to 8u251. You should update to 8u251 (1.8.0_251) or later and see whether that resolves your issue.

**Comment by LiuDui** [comment]
@ajbaldocchi Thanks for your reply. I did run the test for a long time, about a week, but still didn't reproduce the problem.  So I'm not sure if the oracle-jdk has the same problem. I'll update to 8u251 and do some validation. Thank you so much.

**Comment by ruichen984** [comment]
@conicl Do you repeat the problem? I used oracle jdk8u192 and I didn't duplicate it.

**Comment by wForget** [comment]
> hanks for your reply. I did run the test for a long time, about a week, but still didn't reproduce the problem. So I'm not sure if the oracle-jdk has the same problem. I'll update to 8u251 and do some validation. Thank you so much

It is easier to reproduce using Serial GC: ` -XX:+UseSerialGC`
